File: specs/typespec/org/tiers.go
package org

import (
	"errors"
	"slices"

	"vetchium-api-server.typespec/common"
)

var validPlanIDs = []string{"free", "silver", "gold", "enterprise"}

func isValidPlanID(id string) bool {
	return slices.Contains(validPlanIDs, id)
}

// Plan describes a platform plan with its quota caps.
type Plan struct {
	PlanID                 string `json:"plan_id"`
	DisplayName            string `json:"display_name"`
	Description            string `json:"description"`
	DisplayOrder           int32  `json:"display_order"`
	OrgUsersCap            *int32 `json:"org_users_cap,omitempty"`
	DomainsVerifiedCap     *int32 `json:"domains_verified_cap,omitempty"`
	SuborgsCap             *int32 `json:"suborgs_cap,omitempty"`
	MarketplaceListingsCap *int32 `json:"marketplace_listings_cap,omitempty"`
	AuditRetentionDays     *int32 `json:"audit_retention_days,omitempty"`
	SelfUpgradeable        bool   `json:"self_upgradeable"`
}

// PlanUsage holds current usage counts for quota-tracked resources.
type PlanUsage struct {
	OrgUsers            int32 `json:"org_users"`
	DomainsVerified     int32 `json:"domains_verified"`
	Suborgs             int32 `json:"suborgs"`
	MarketplaceListings int32 `json:"marketplace_listings"`
}

// OrgPlan is the full subscription record with plan and usage.
type OrgPlan struct {
	OrgID       string    `json:"org_id"`
	OrgDomain   string    `json:"org_domain"`
	CurrentPlan Plan      `json:"current_plan"`
	Usage       PlanUsage `json:"usage"`
	UpdatedAt   string    `json:"updated_at"`
	Note        string    `json:"note"`
}

// ListPlansRequest is the request to list all active plans.
type ListPlansRequest struct{}

func (r ListPlansRequest) Validate() []common.ValidationError { return nil }

// ListPlansResponse wraps the plans list.
type ListPlansResponse struct {
	Plans []Plan `json:"plans"`
}

// GetMyOrgPlanRequest is the request to get the caller's org plan.
type GetMyOrgPlanRequest struct{}

func (r GetMyOrgPlanRequest) Validate() []common.ValidationError { return nil }

// UpgradeOrgPlanRequest upgrades to the given plan.
type UpgradeOrgPlanRequest struct {
	PlanID string `json:"plan_id"`
}

func (r UpgradeOrgPlanRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if !isValidPlanID(r.PlanID) {
		errs = append(errs, common.NewValidationError("plan_id", errInvalidPlanID))
	}
	return errs
}

// AdminListOrgPlansRequest lists all org plans (admin).
type AdminListOrgPlansRequest struct {
	FilterPlanID  *string `json:"filter_plan_id,omitempty"`
	FilterDomain  *string `json:"filter_domain,omitempty"`
	PaginationKey *string `json:"pagination_key,omitempty"`
	Limit         *int32  `json:"limit,omitempty"`
}

func (r AdminListOrgPlansRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.FilterPlanID != nil && *r.FilterPlanID != "" && !isValidPlanID(*r.FilterPlanID) {
		errs = append(errs, common.NewValidationError("filter_plan_id", errInvalidPlanID))
	}
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.NewValidationError("limit", errLimitOutOfRange))
	}
	return errs
}

// AdminListOrgPlansResponse wraps the plan list.
type AdminListOrgPlansResponse struct {
	Items             []OrgPlan `json:"items"`
	NextPaginationKey *string   `json:"next_pagination_key,omitempty"`
}

// AdminSetOrgPlanRequest sets the plan for a given org (admin).
type AdminSetOrgPlanRequest struct {
	OrgID  string `json:"org_id"`
	PlanID string `json:"plan_id"`
	Reason string `json:"reason"`
}

func (r AdminSetOrgPlanRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.OrgID == "" {
		errs = append(errs, common.NewValidationError("org_id", errOrgIDRequired))
	}
	if !isValidPlanID(r.PlanID) {
		errs = append(errs, common.NewValidationError("plan_id", errInvalidPlanID))
	}
	if r.Reason == "" {
		errs = append(errs, common.NewValidationError("reason", errReasonRequired))
	} else if len(r.Reason) > 2000 {
		errs = append(errs, common.NewValidationError("reason", errReasonTooLong))
	}
	return errs
}

var (
	errInvalidPlanID   = errors.New("must be a valid plan id (free, silver, gold, enterprise)")
	errLimitOutOfRange = errors.New("must be between 1 and 100")
	errOrgIDRequired   = errors.New("org_id is required")
	errReasonRequired  = errors.New("reason is required")
	errReasonTooLong   = errors.New("must be at most 2000 characters")
)

File: specs/typespec/org/tiers.ts
import { newValidationError, type ValidationError } from "../common/common";

export interface OrgTier {
	tier_id: string;
	display_name: string;
	description: string;
	display_order: number;
	org_users_cap?: number;
	domains_verified_cap?: number;
	suborgs_cap?: number;
	marketplace_listings_cap?: number;
	audit_retention_days?: number;
	self_upgradeable: boolean;
}

export interface OrgTierUsage {
	org_users: number;
	domains_verified: number;
	suborgs: number;
	marketplace_listings: number;
}

export interface OrgSubscription {
	org_id: string;
	org_domain: string;
	current_tier: OrgTier;
	usage: OrgTierUsage;
	updated_at: string;
	note: string;
}

export interface ListOrgTiersRequest {}

export interface ListOrgTiersResponse {
	tiers: OrgTier[];
}

export interface GetMyOrgSubscriptionRequest {}

export interface SelfUpgradeOrgSubscriptionRequest {
	tier_id: string;
}

export interface AdminListOrgSubscriptionsRequest {
	filter_tier_id?: string;
	pagination_key?: string;
	limit?: number;
}

export interface AdminListOrgSubscriptionsResponse {
	items: OrgSubscription[];
	next_pagination_key?: string;
}

export interface AdminSetOrgTierRequest {
	org_id: string;
	tier_id: string;
	reason: string;
}

const VALID_TIER_IDS = ["free", "silver", "gold", "enterprise"] as const;
type TierId = (typeof VALID_TIER_IDS)[number];

export function validateSelfUpgradeOrgSubscriptionRequest(
	req: SelfUpgradeOrgSubscriptionRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!req.tier_id || !VALID_TIER_IDS.includes(req.tier_id as TierId)) {
		errs.push(newValidationError("tier_id", "must be a valid tier id"));
	}
	return errs;
}

export function validateAdminListOrgSubscriptionsRequest(
	req: AdminListOrgSubscriptionsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (
		req.filter_tier_id !== undefined &&
		req.filter_tier_id !== "" &&
		!VALID_TIER_IDS.includes(req.filter_tier_id as TierId)
	) {
		errs.push(newValidationError("filter_tier_id", "must be a valid tier id"));
	}
	if (req.limit !== undefined && (req.limit < 1 || req.limit > 100)) {
		errs.push(newValidationError("limit", "must be between 1 and 100"));
	}
	return errs;
}

export function validateAdminSetOrgTierRequest(
	req: AdminSetOrgTierRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!req.org_id) {
		errs.push(newValidationError("org_id", "org_id is required"));
	}
	if (!req.tier_id || !VALID_TIER_IDS.includes(req.tier_id as TierId)) {
		errs.push(newValidationError("tier_id", "must be a valid tier id"));
	}
	if (req.reason === undefined || req.reason === null) {
		errs.push(newValidationError("reason", "reason is required"));
	} else if (req.reason.length > 2000) {
		errs.push(newValidationError("reason", "must be at most 2000 characters"));
	}
	return errs;
}
